algorithms/DFS/dfs.cpp

```diff
@@ -0,0 +1,136 @@
+#include <map>
+#include <stdlib.h>
+#include <vector>
+#include <iostream>
+#include <unordered_set>
+#include <unordered_map>
+#include <stack>
+
+namespace DFS
+{
+	using AdjacencyList = std::map<size_t, std::vector<size_t>>;
+	using NodeList = std::vector<size_t>;
+	class DFS
+	{
+	public:
+		DFS(AdjacencyList aList, bool verbose) : mE(aList), mVerbose(verbose)
+		{
+			for (auto it = mE.begin(); it != mE.end(); it++)
+			{
+				mV.push_back(it->first);
+			}
+			mComponent.resize(mV.size());
+		}
+
+		void process()
+		{
+			for (size_t v : mV)
+			{
+				if (mVisited.contains(v))
+				{
+					continue;
+				}
+				mVisited.insert(v);
+				mRoot(v);
+				mDfs(v, v);
+			}
+		}
+
+		std::vector<size_t> getComponent()
+		{
+			return mComponent;
+		}
+
+		size_t getUniqueComponents()
+		{
+			return std::unordered_set<size_t>(mComponent.begin(), mComponent.end()).size();
+		}
+
+	private:
+		void mRoot(size_t v)
+		{
+			mOReps.push(v);
+			mONodes.push_back(v);
+			mDfsNum.insert_or_assign(v, mDfsPos++);
+		}
+
+		void mDfs(size_t u, size_t v)
+		{
+			if (mVerbose)
+				std::cout << "DFS("<<u << ", " << v << ")\n"; 
+
+			for (size_t w : mE.at(v))
+			{
+				if (mVisited.contains(w))
+				{
+					mTraverseNonEdgeTree(v, w);
+					continue;
+				}
+				mTraverseTreeEdge(v, w);
+				mVisited.insert(w);
+				mDfs(v, w);
+			}
+			mBacktrack(u, v);
+		}
+		void mTraverseNonEdgeTree(size_t v, size_t w)
+		{
+			if (mVerbose)
+				std::cout << "TRAVERSE_NONTREEEDGE("<<v << ", " << w << ")\n"; 
+			for (size_t x : mONodes)
+			{
+				if( x == w)
+				{
+					while (mDfsNum[w] < mDfsNum[mOReps.top()])
+					{
+						mOReps.pop();
+					}
+					break;
+				}
+			}
```

Comment:
This seems too complex for just popping all values in mDfsNum with value larger than w. To make sure w i in mONodes maybe std::find could help solving this.

Reply:
I just copied the code from the lecture, idk

Reply:
Your implementation definitely works aswell, but I replaced it with the std::find this looks a bit better :D